base_export_async/models/delay_export.py

```diff
@@ -0,0 +1,119 @@
+class DelayExport(models.Model):
+
+    _name = 'delay.export'
+    _description = 'Allow to delay the export'
+
+    user_id = fields.Many2one('res.users', string='User', index=True)
+
+    @api.model
+    def delay_export(self, data):
+        params = json.loads(data.get('data'))
+        if not self.env.user.email:
+            raise UserError(_("You must set an email address to your user."))
+        self.with_delay().export(params)
+
+    @api.model
+    @job
+    def export(self, params):
+        export_format = params.get('format')
+        raw_data = export_format != 'csv'
+
+        model_name, fields_name, ids, domain, import_compat, context = \
+            operator.itemgetter('model', 'fields', 'ids',
+                                'domain', 'import_compat', 'context')(params)
+        user = self.env['res.users'].browse([context.get('uid')])
+        if not user or not user.email:
+            raise UserError(_("The user doesn't have an email address."))
+
+        model = self.env[model_name].with_context(
+            import_compat=import_compat, **context)
+        records = model.browse(ids) or model.search(
+            domain, offset=0, limit=False, order=False)
+
+        if not model._is_an_ordinary_table():
+            fields_name = [field for field in fields_name
+                           if field['name'] != 'id']
+
+        field_names = [f['name'] for f in fields_name]
+        import_data = records.export_data(
+            field_names, raw_data).get('datas', [])
+
+        if import_compat:
+            columns_headers = field_names
+        else:
+            columns_headers = [val['label'].strip() for val in fields_name]
+
+        if export_format == 'csv':
+            csv = CSVExport()
+            result = csv.from_data(columns_headers, import_data)
+        else:
+            xls = ExcelExport()
+            result = xls.from_data(columns_headers, import_data)
+
+        export_record = self.sudo().create({'user_id': user.id})
+
+        name = "{}.{}".format(model_name, export_format)
+        attachment = self.env['ir.attachment'].create({
+            'name': name,
+            'datas': base64.b64encode(result),
+            'datas_fname': name,
+            'type': 'binary',
+            'res_model': self._name,
+            'res_id': export_record.id,
+        })
+
+        url = "{}/web/content/ir.attachment/{}/datas/{}?download=true".format(
+            self.env['ir.config_parameter'].sudo().get_param('web.base.url'),
+            attachment.id,
+            attachment.name,
+        )
+
+        time_to_live = self.env['ir.config_parameter'].sudo(). \
+            get_param('attachment.time.to.live', 7)
+        date_today = fields.Date.today()
+        expiration_date = fields.Date.to_string(
+            date_today + relativedelta(days=+int(time_to_live)))
+
+        odoo_bot = self.sudo().env.ref("base.partner_root")
+        email_from = odoo_bot.email
+        self.env['mail.mail'].create({
```

Comment:
I'd add `# TODO: move to mail template`

Reply:
I don't like to add #TODO. Either we decide to add a mail template (Which is a good Idea) either we let it this way.